src/xr/xr-hit-test.js

```diff
@@ -0,0 +1,324 @@
+            if (fireError) fireError.fire('error', err);
+            return false;
+        }
+
+        return true;
+    };
+
+    /**
+     * @function
+     * @name pc.XrHitTest#start
+     * @description Attempts to start hit test with provided reference space.
+     * @param {string} [spaceType] - Optional reference space type. Defaults to {pc.XRSPACE_VIEWER}. Can be one of the following:
+     *
+     * * {@link pc.XRSPACE_VIEWER}: Viewer - hit test will be facing relative to viewers space.
+     * * {@link pc.XRSPACE_LOCAL}: Local - represents a tracking space with a native origin near the viewer at the time of creation.
+     * * {@link pc.XRSPACE_LOCALFLOOR}: Local Floor - represents a tracking space with a native origin at the floor in a safe position for the user to stand. The y axis equals 0 at floor level. Floor level value might be estimated by the underlying platform.
+     * * {@link pc.XRSPACE_BOUNDEDFLOOR}: Bounded Floor - represents a tracking space with its native origin at the floor, where the user is expected to move within a pre-established boundary.
+     * * {@link pc.XRSPACE_UNBOUNDED}: Unbounded - represents a tracking space where the user is expected to move freely around their environment, potentially long distances from their starting point.
+     *
+     * @param {string[]} [entityTypes] - Optional list of underlying entity types against which hit tests will be performed. Defaults to [ {pc.XRTRACKABLE_PLANE} ]. Can be any combination of the following:
+     *
+     * * {@link pc.XRTRACKABLE_POINT}: Point - indicates that the hit test results will be computed based on the feature points detected by the underlying Augmented Reality system.
+     * * {@link pc.XRTRACKABLE_PLANE}: Plane - indicates that the hit test results will be computed based on the planes detected by the underlying Augmented Reality system.
+     * * {@link pc.XRTRACKABLE_MESH}: Mesh - indicates that the hit test results will be computed based on the meshes detected by the underlying Augmented Reality system.
+     *
+     * @param {pc.Ray} [offsetRay] - Optional ray by which hit test ray can be offset.
+     * @param {pc.callbacks.XrHitTestStart} [callback] - Optional callback function called once hit test source is created or failed.
+     * @example
+     * app.xr.hitTest.start(function (err, hitTestSource) {
+     *     if (err) return;
+     *     hitTestSource.on('result', function (position, rotation) {
+     *         // position and rotation of hit test result
+     *         // based on default Ray facing forward from the Viewer (default space)
+     *     });
+     * });
+     * @example
+     * var ray = new pc.Ray(new pc.Vec3(0, 0, 0), new pc.Vec3(0, -1, 0));
+     * app.xr.hitTest.start(pc.XRSPACE_LOCAL, [pc.XRTRACKABLE_PLANE], ray, function (err, hitTestSource) {
+     *     // hit test source that will sample real world geometry straight down
+     *     // from the position where AR session started
+     * });
```

Comment:
I would rather there weren't multiple function signatures (it makes the TypeScript definitions wrong and probably makes it difficult for the JIT to optimize the code). Can we break this out into multiple functions with fixed parameter lists?